### PR TITLE
Implement W3C Trace Context propagation.

### PR DESCRIPTION
**Pull request by gtchiflidjanov**
- Make Trace and Span ID generation configurable and add a generator for W3C ID size.
- Add configurable TextMap and HTTPHeaders propagator which Inject/Extract SpanContext to/from W3C traceparent/tracestate headers.

### CODE REVIEW

**Review by akhani18 — APPROVED** [review]
Thanks for making this change.

File: tracer/w3c.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		}
+		if _, err := hex.DecodeString(parts[i]); err != nil {
+			return opentracing.ErrSpanContextCorrupted
+		}
+	}
+	if parts[0] != "00" {
+		return opentracing.ErrSpanContextCorrupted
+	}
+
+	sc.TraceID, _ = ToUUID(parts[1])
```
[the remark this answers is not in the document]
Comment (gtchiflidjanov) [reply tracer/w3c.go:108]: In `jaeger_wavefront.go`

File: tracer/w3c.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+		case tracestate:
+			return tracestateParse(v, &sc)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(sc.TraceID) == 0 {
```
[the remark this answers is not in the document]
Comment (gtchiflidjanov) [reply tracer/w3c.go:72]: The `traceparentParse` sets both TraceID and SpanID or return an error. This checks if it was invoked.

**Review by hanwavefront — APPROVED** [review]
Thanks! LGTM!

File: tracer/tracer.go [hunk 3]
```diff
@@ -62,10 +71,25 @@ func WithJaegerPropagator(traceId, baggagePrefix string) Option {
...
+	return func(t *WavefrontTracer) {
+		t.generator = NewGeneratorW3C()
+	}
+}
+
+func WithW3CPropagator() Option {
+	return func(t *WavefrontTracer) {
+		// implies W3C Generator
+		t.generator = NewGeneratorW3C()
+		t.textPropagator = NewW3CPropagator()
```
Comment (thepeterstone) [inline tracer/tracer.go:84]: Maybe rename this to `NewPropagatorW3C` following the convention from the generators?

File: tracer/w3c.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+const (
+	traceparent = "traceparent"
+	tracestate  = "tracestate"
+)
+
+type W3CPropagator struct {
+}
+
+func NewW3CPropagator() SpanContextPropagator {
```
Comment (thepeterstone) [inline tracer/w3c.go:19]: It would be great to add some godoc comments to the new code

**Review by thepeterstone — APPROVED** [review]
A few minor style concerns, otherwise looks good to me

**Review by thepeterstone — APPROVED** [review]

**Pull request merged**